lib/Install.js

```diff
@@ -1,86 +1,71 @@
     // someone ran `elm-test install`. We do not delete this directory after
     // the installation finishes in case the user needs to debug the test run.
-    if (fs.existsSync(dirPath)) {
+    if (fs.existsSync(generatedCodeDir)) {
       // We can replace this with `fs.rmdirSync(dir, { recursive: true })`
       // once Node.js 10 is EOL 2021-04-30 and support for Node.js 10 is dropped.
-      rimraf.sync(dirPath);
+      rimraf.sync(generatedCodeDir);
     }
-    fs.mkdirSync(dirPath, { recursive: true });
-  } catch (err) {
-    console.error(
-      'Unable to create temporary directory for elm-test install.',
-      err
+    fs.mkdirSync(generatedCodeDir, { recursive: true });
+  } catch (error) {
+    throw new Error(
+      `Unable to create temporary directory for elm-test install: ${error.message}`
     );
-    process.exit(1);
   }
 
-  var elmJson = JSON.parse(fs.readFileSync('elm.json', 'utf8'));
-  var tmpElmJsonPath = path.join(dirPath, 'elm.json');
-  var isPackage;
-
-  switch (elmJson['type']) {
-    case 'package':
-      isPackage = true;
-      break;
+  const { elmJson } = project;
 
-    case 'application':
-      isPackage = false;
-      break;
-
-    default:
-      console.error('Unrecognized elm.json type:', elmJson['type']);
-      process.exit(1);
-  }
-
-  // This mirrors the behavior of `elm install` passing a package that is
-  // already installed. Say it's already installed, then exit 0.
   if (
-    (isPackage && elmJson['test-dependencies'].hasOwnProperty(packageName)) ||
-    (!isPackage &&
-      elmJson['test-dependencies']['direct'].hasOwnProperty(packageName))
+    elmJson.type === 'package'
+      ? elmJson['test-dependencies'].hasOwnProperty(packageName)
+      : elmJson['test-dependencies'].direct.hasOwnProperty(packageName)
   ) {
-    console.log('It is already installed!');
-    return;
+    return 'AlreadyInstalled';
   }
 
-  oldSourceDirectories = elmJson['source-directories'];
-
-  // Without this, `elm install` will complain about missing source dirs
-  // in the temp dir. This way we don't have to create them!
-  elmJson['source-directories'] = ['.'];
-
-  fs.writeFileSync(tmpElmJsonPath, JSON.stringify(elmJson), 'utf8');
-
-  try {
-    child_process.execFileSync(pathToElmBinary, ['install', packageName], {
-      stdio: 'inherit',
-      cwd: dirPath,
-    });
-  } catch (error) {
-    process.exit(error.status || 1);
+  const tmpElmJson =
+    elmJson.type === 'package'
+      ? elmJson
+      : {
+          ...elmJson,
+          // Without this, `elm install` will complain about missing source dirs
+          // in the temp dir. This way we don't have to create them!
+          'source-directories': ['.'],
+        };
+
+  ElmJson.write(generatedCodeDir, tmpElmJson);
+
+  const result = spawn.sync(pathToElmBinary, ['install', packageName], {
+    stdio: 'inherit',
+    cwd: generatedCodeDir,
+  });
+
+  if (result.status !== 0) {
+    process.exit(result.status);
```

Comment:
I tried to concentrate all `process.exit` in elm-test.js, but I couldn’t figure out a good way to do it here. In this case we have run `elm install` and printed its output to the console. If `elm install` failed we just want to exit with `elm install`’s error output and not print anything else or so. I think this is fine.

Reply:
:+1: